Loader full reconciliation

dylandreimerink opened the pull request:
WIP

dylandreimerink committed:
WIP

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki committed:
loader/orchestrator: WIP reworking

- Remove datapath.Loader and only use loader/types.Loader
- Add LoaderContext for passing device and address data to the loader.
  In the future this can contain other data queried from the agent.
- Query for node addresses in the orchestrator instead of loader and
  pass them as part of LoaderContext.
- Add rate limiting so we don't reinit twice due to propagation delay
  to Table[NodeAddress]

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/datapath/loader/types/types.go:
```diff
@@ -5,25 +5,41 @@ package types
 )
 
+// LoaderContext are the external inputs to the loader resolved by the orchestrator.
+type LoaderContext struct {
```
This was added by me and I'm not quite sure about the name. The general idea with this is to make the loader "self-sustained" and have all inputs to it are passed as parameters to its methods, with the orchestrator dealing with where to get the data and how to sequence different operations, so that loader is left with dealing with the actual mechanics. 

I'm sure calling it "context" would be confused with `context.Context`. Calling it `State` isn't quite right either, as it really is the "context" in which the loader operation happens (could also put `context.Context` into this?)

joamaki committed:
loader: Explicitly pass address info to ConfigWriter

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
orchestrator: Add simple retry mechanism

Retry failed reinitialization after 10 second delay.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loader: Only pass relevant data from local node in LoaderContext

If we pass the whole LocalNode we will reinitialize on irrelevant
changes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
various fixes

- drop useless ipsecMu
- pass LoaderContext in template cache update to recompute hash based on it
- also collect the fallback node addresses

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
tables: Remove fallback node addresses

This mechanism proved out to make the use of the Table[NodeAddress]
inconvenient. Reimplement the fallback as a computation around a set
of node addresses instead.

It also was not useful to have the fallback addresses from non-selected
devices.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Remove unnecessary Chdir() that caused a race

The Chdir() done by Daemon.init was racing with the loader causing
just compiled object files to not be found. Change the Chdir() in
initEnv() to switch to StateDir directly and remove the Chdir() from
Daemon.init.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
orchestrator: Add logging

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loader: Oops, use the internal IP not node IP for cilium_host

Signed-off-by: Jussi Maki <jussi@isovalent.com>
